libbpf-cargo: Relative anonymous type naming for struct fields

Waujito opened the pull request:
Previously, anonymous types were named as __anon_{id} with id as a global counter. This approach caused cascading renaming of all subsequent anonymous definitions after any C anonymous struct order/amount change. This was a major design problem not for only development but even portability, since here may be the case when something was changed in vmlinux.h leading to compilation errors.

This PR addresses the issue by localizing cascading effects to the local struct layer rather than the global layer. While the solution is not fully ideal - cascading still occurs within individual structs - it eliminates the need for wide Rust code modifications when the order of anonymous struct definitions changes.

The PR is breaking and should be included in major version update.

The PR fixes #1161

While this approach is ok, I still think the better way will be to use getters, since this is breaking and less intuitive.

danielocfb commented:
Haven't looked at it yet, but this is great. I was considering doing that for `0.25.0`, but didn't have the cycles. Will postpone the release until I've gotten a chance to take a look here. Please look into the CI failures, which all seem legit.

Waujito commented:
Here are still some plain __anon types like in `maps`. I'll try to catch more and may be add some defenders of name collisions.

danielocfb commented on libbpf-cargo/src/gen/btf.rs:
```diff
@@ -377,15 +385,70 @@ pub(crate) struct TypeMap {
+            if let Some(n) = parent.name() {
+                dep.name = Some(n.to_string_lossy().to_string());
+            }
+            if ty.name().is_some() {
+                dep.child_counter = Rc::new(RefCell::new(0));
+            }
```
Why is this code needed?

Waujito replied:
Added a comment

danielocfb commented on libbpf-cargo/src/gen/btf.rs:
```diff
@@ -377,15 +385,70 @@ pub(crate) struct TypeMap {
+            if ty.name().is_some() {
+                dep.child_counter = Rc::new(RefCell::new(0));
+            }
+            dep.dep_id = 1;
```
Why is this value set here and not already during initialization?

Waujito replied:
Added a comment

danielocfb commented on libbpf-cargo/src/gen/btf.rs:
```diff
@@ -377,15 +385,70 @@ pub(crate) struct TypeMap {
+            };
+            deps.insert(parent.type_id(), dep.clone());
+
+            if let Some(n) = parent.name() {
+                dep.name = Some(n.to_string_lossy().to_string());
+            }
```
Why is this value set here and not during initialization?

danielocfb commented on libbpf-cargo/src/gen/btf.rs:
```diff
@@ -377,15 +385,70 @@ pub(crate) struct TypeMap {
 }
 
 impl TypeMap {
+    pub fn derive_parent<'s>(&self, ty: &BtfType<'s>, parent: &BtfType<'s>) {
```
Is the name really fitting? Aren't you more registering a parent relationship rather than "deriving" anything?

Waujito replied:
assign_parent?

Waujito replied:
Done

danielocfb commented on libbpf-cargo/src/gen/btf.rs:
```diff
@@ -367,6 +368,13 @@ fn escape_reserved_keyword(identifier: Cow<'_, str>) -> Cow<'_, str> {
 
+#[derive(Debug, Clone)]
+pub struct BtfDependency {
+    pub name: Option<String>,
+    pub dep_id: i32,
+    pub child_counter: Rc<RefCell<i32>>,
```
Please document what these represent.

danielocfb reviewed (commented):
In general this seems okay to me, but the code seems under documented and under-tested in the sense that I can remove/change various lines and all tests keep passing. Similarly, I am not sure I understand some of the assignments done. Left some comments.

danielocfb commented:
> Here are still some plain __anon types like in `maps`. I'll try to catch more and may be add some defenders of name collisions.

What do you mean exactly? Where is "here"?

Waujito commented:
In [0e504e6](https://github.com/libbpf/libbpf-rs/pull/1178/commits/0e604e6b803c68dc613613657247c0129269637c) I simplified the code slightly, modified the global anon counting behavior, added documentation for new structure fields and included a test for depth-aligned anonymous structs.

The global anonymous struct counting now excludes relative anons to make its behavior more representative and easier to use. I also added some fallback protection on it for (im)possible case when global anonymous struct is assigned before a relative one: in such case we will use old type of naming.


> > Here are still some plain __anon types like in `maps`. I'll try to catch more and may be add some defenders of name collisions.
> 
> What do you mean exactly? Where is "here"?

Nvm. Those anonymous structs are at the global level and should be kept with old naming.

I think the code is ok now, but I'm aware of possible name collisions for relative types. 
I can't mind the case, but *what if*. Maybe I should implement dep id counting in a separate hashmap for parent names as strings, but not with that crazy `child_counter`? This solution will not be elegant, but more trustworthy...

danielocfb commented:
Thanks, it looks better now!

> I think the code is ok now, but I'm aware of possible name collisions for relative types.
> I can't mind the case, but _what if_.

Can you provide an example of such a case?

Waujito commented:
> Can you provide an example of such a case?

I'm almost sure here is no the case. I was just thinking about it
